pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java

```diff
@@ -296,16 +299,31 @@ public MetadataStoreExtended createConfigurationMetadataStore() throws MetadataS
                         .build());
     }
 
+    @Override
+    public void close() throws PulsarServerException {
+        try {
+            closeAsync().get();
+        } catch (PulsarServerException e) {
+            throw e;
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof PulsarServerException) {
+                throw (PulsarServerException) e.getCause();
+            } else {
+                throw new PulsarServerException(e.getCause());
+            }
+        } catch (Exception e) {
+            throw new PulsarServerException(e);
+        }
+    }
+
     /**
      * Close the current pulsar service. All resources are released.
      */
-    @Override
-    public void close() throws PulsarServerException {
+    public CompletableFuture<Void> closeAsync() throws PulsarServerException {
```

Comment:
generally it is a code smell to see a method that returns a CompletableFuture to throw an exception.

The expectations for the caller is that any error will be reported by the CompletableFuture.

can we catch all of the possible exceptions and then return a failed CompletableFuture ?

Reply:
In this case making that change wouldn't make an actual difference since `closeAsync` isn't part of a public API exposed outside of Pulsar.

Reply:
It's possible to make that change, but is it necessary?

Reply:
for instance this code looks bad
https://github.com/apache/pulsar/pull/9308/files#diff-be597a630f836a289aa83fa608ae27fd751e5b94577fc59701ee9e9ab9fd8be3R69

Reply:
> for instance this code looks bad

What specifically is "bad" there?

Reply:
resolved in recent changes